umh-core/pkg/communicator/actions/stateMocker.go

```diff
@@ -0,0 +1,252 @@
+	return s.ConfigManager
+}
+
+// Tick advances the state mocker by one tick and updates the system state
+func (s *StateMocker) Tick() {
+	s.TickCounter++
+	s.UpdateState()
+}
+
+// here, the actual state update logic is implemented
+func (s *StateMocker) UpdateState() {
+	// only take the dataflowcomponent configs and add them to the state of the dataflowcomponent manager
+	// the other managers are not updated
+
+	//start with a basic snapshot
+	dfcManagerInstaces := map[string]*fsm.FSMInstanceSnapshot{}
+
+	for _, curDataflowcomponent := range s.ConfigManager.GetDataFlowConfig() {
+		// get the default currentState from the last observed state (s.state)
+		currentState := curDataflowcomponent.DesiredFSMState
+		var instances map[string]*fsm.FSMInstanceSnapshot
+		if s.State != nil && s.State.Managers != nil {
+			if manager, ok := s.State.Managers[constants.DataflowcomponentManagerName]; ok {
+				instances = manager.GetInstances()
+			}
+		}
+
+		if instance, ok := instances[curDataflowcomponent.Name]; ok {
+			currentState = instance.CurrentState
+		}
+
+		// Check if there are pending transitions for this component
+		if transitions, exists := s.PendingTransitions[curDataflowcomponent.Name]; exists {
+			// Find transitions that should be applied at this tick
+			for i, transition := range transitions {
+				if transition.TickAt <= s.TickCounter {
+					currentState = transition.State
+
+					// Remove applied transitions if they've been processed
+					if i < len(transitions)-1 {
+						s.PendingTransitions[curDataflowcomponent.Name] = transitions[i+1:]
+					} else {
+						// All transitions applied, clear the list
+						delete(s.PendingTransitions, curDataflowcomponent.Name)
+					}
+					break
+				}
+			}
+		}
+
+		dfcManagerInstaces[curDataflowcomponent.Name] = &fsm.FSMInstanceSnapshot{
+			ID:           curDataflowcomponent.Name,
+			DesiredState: curDataflowcomponent.DesiredFSMState,
+			CurrentState: currentState,
+			LastObservedState: &dataflowcomponent.DataflowComponentObservedStateSnapshot{
+				Config: curDataflowcomponent.DataFlowComponentServiceConfig,
+			},
+		}
+	}
+
+	managerSnapshot := &MockManagerSnapshot{
+		Instances: dfcManagerInstaces,
+	}
+
+	// update the state of the system with the new manager snapshot
+	if s.State.Managers == nil {
+		s.State.Managers = make(map[string]fsm.ManagerSnapshot)
+	}
+	s.State.Managers[constants.DataflowcomponentManagerName] = managerSnapshot
+}
+
+// Start spawns a goroutine that updates the state of the system periodically
+// It returns immediately and can be stopped by calling Stop()
+func (s *StateMocker) Run() {
+	ticker := time.NewTicker(1 * time.Second)
+
+	go func() {
+		defer ticker.Stop()
+
+		for {
+			select {
+			case <-ticker.C:
+				s.Tick()
+			case <-s.done:
+				return
+			}
+		}
+	}()
+}
+
+// Start is a convenience method that runs the state mocker in a new goroutine
+func (s *StateMocker) Start() {
+	go s.Run()
```

Comment:
For safety it could make sense to throw an error if this is started twice without stopping first.